lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -0,0 +1,481 @@
+  coerce @(IO (Maybe (Sqlite.Only a))) @(IO (Maybe a)) (queryMaybe_ conn s)
+
+queryOne_ :: Sqlite.FromRow a => Connection -> Sql -> IO a
+queryOne_ conn s =
+  queryListCheck_ conn s \case
+    [x] -> Right x
+    xs -> Left (SomeShowTypeable (ExpectedExactlyOneRowException (anythingToString xs)))
+
+queryOneOne_ :: forall a. Sqlite.FromField a => Connection -> Sql -> IO a
+queryOneOne_ conn s =
+  coerce @(IO (Sqlite.Only a)) @(IO a) (queryOne_ conn s)
+
+-- With results, without parameters, with checks
+
+queryListCheck_ :: (Sqlite.FromRow a, Show e, Typeable e) => Connection -> Sql -> ([a] -> Either e r) -> IO r
+queryListCheck_ conn s check =
+  gqueryListCheck_ conn s (mapLeft SomeShowTypeable . check)
+
+gqueryListCheck_ :: Sqlite.FromRow a => Connection -> Sql -> ([a] -> Either SomeShowTypeable r) -> IO r
+gqueryListCheck_ conn s check = do
+  xs <- queryList_ conn s
+  case check xs of
+    Left exception ->
+      throwSqliteException
+        SqliteExceptionInfo
+          { connection = conn,
+            exception,
+            params = Nothing,
+            sql = s
+          }
+    Right result -> pure result
+
+queryListOneCheck_ ::
+  forall a e r.
+  (Sqlite.FromField a, Show e, Typeable e) =>
+  Connection ->
+  Sql ->
+  ([a] -> Either e r) ->
+  IO r
+queryListOneCheck_ conn s check =
+  queryListCheck_ conn s (coerce @([a] -> Either e r) @([Sqlite.Only a] -> Either e r) check)
+
+queryMaybeCheck_ :: (Sqlite.FromRow a, Show e, Typeable e) => Connection -> Sql -> (Maybe a -> Either e r) -> IO r
+queryMaybeCheck_ conn s check =
+  gqueryListCheck_ conn s \case
+    [] -> mapLeft SomeShowTypeable (check Nothing)
+    [x] -> mapLeft SomeShowTypeable (check (Just x))
+    xs -> Left (SomeShowTypeable (ExpectedAtMostOneRowException (anythingToString xs)))
+
+queryMaybeOneCheck_ ::
+  forall a e r.
+  (Sqlite.FromField a, Show e, Typeable e) =>
+  Connection ->
+  Sql ->
+  (Maybe a -> Either e r) ->
+  IO r
+queryMaybeOneCheck_ conn s check =
+  queryMaybeCheck_ conn s (coerce @(Maybe a -> Either e r) @(Maybe (Sqlite.Only a) -> Either e r) check)
+
+queryOneCheck_ :: (Sqlite.FromRow a, Show e, Typeable e) => Connection -> Sql -> (a -> Either e r) -> IO r
+queryOneCheck_ conn s check =
+  gqueryListCheck_ conn s \case
+    [x] -> mapLeft SomeShowTypeable (check x)
+    xs -> Left (SomeShowTypeable (ExpectedExactlyOneRowException (anythingToString xs)))
+
+queryOneOneCheck_ ::
+  forall a e r.
+  (Sqlite.FromField a, Show e, Typeable e) =>
+  Connection ->
+  Sql ->
+  (a -> Either e r) ->
+  IO r
+queryOneOneCheck_ conn s check =
+  queryOneCheck_ conn s (coerce @(a -> Either e r) @(Sqlite.Only a -> Either e r) check)
+
+-- Low-level
+
+-- | Perform an action within a named savepoint. The action is provided a rollback action.
+withSavepoint :: Connection -> Text -> (IO () -> IO a) -> IO a
+withSavepoint conn name action =
+  bracket_
+    (execute_ conn (Sql ("SAVEPOINT " <> name)))
+    (execute_ conn (Sql ("RELEASE " <> name)))
```

Comment:
I think we should automatically roll-back if we hit an exception (you can add a rollback to the user action using `onException`); otherwise this will commit a partial transaction if an exception is encountered.

Reply:
Good _catch_